### ISSUE TITLE
about bias quantization

### ISSUE THREAD

**Issue by sonebu** [MaximIntegratedAI/ai8x-synthesis#123]
When generating "deployment code" for networks that have layer weights with bitwidths different than 8, the corresponding bias values for those layers seem to get constrained to the same number of steps as the weights. For instance, a layer with 8-bit weights gets "8-bit" bias values (i.e., there are 256 different possible choices for each bias element from the dictionary {0x00, 0x01, ... 0xfe, 0xff}), but a layer with 2-bit weights gets only 4 different bias value possibilities, i.e., can only use the dictionary {0x00, 0x01, 0xfe, 0xff}, which effectively makes it 2-bit.

I've noticed this while generating code for the mixed-precision CIFAR-100 simplenet checkpoint (i.e., "[ai85-cifar100-qat-mixed-q.pth.tar](https://github.com/MaximIntegratedAI/ai8x-synthesis/blob/bb712b97e8f27d24a51a787a3b78125460da52e1/trained/ai85-cifar100-qat-mixed-q.pth.tar)"). Steps to reproduce:

1. clone+setup ai8x-synthesis, checkout bb712b97e8f27d24a51a787a3b78125460da52e1  (latest development commit AFAIK, Apr 16 2:02 AM GMT+3)
2. generate code for "ai85-cifar100-qat-mixed-q.pth.tar":
`./ai8xize.py --verbose --log --test-dir sdk/Examples/MAX78000/CNN --prefix cifar-100-mixed --checkpoint-file trained/ai85-cifar100-qat-mixed-q.pth.tar --config-file networks/cifar100-simple.yaml --softmax --device MAX78000 --compact-data --mexpress --timer 0 --display-checkpoint --boost 2.5`
3. check file sdk/Examples/MAX78000/CNN/cifar-100-mixed/weights.h for the arrays "BIAS_0", 1, 2 and 3. Bias values corresponding to the 2-bit layers only take the values {0xff, 0xfe, 0x00, 0x01}, which correspond respectively to {-2,-1,0,+1}, and the case is similar for the bias values corresponding to 4-bit-weight layers (16 values there). I identified the indices of the values corresponding to 2-bit and 4-bit layers by explicitly setting various elements to 0 in the state_dict and regenerating the code, but even just looking at the bare array with the real values actually hints at what I'm trying to say. 

I think this behavior is caused by the following: In quantize.py, [bias values are left shifted by 7](https://github.com/MaximIntegratedAI/ai8x-synthesis/blob/bb712b97e8f27d24a51a787a3b78125460da52e1/izer/quantize.py#L223) so that "PyTorch can still use them to run a model.". The comment also states that "This needs to be reversed before loading the weights into the hardware."; and this is done -> When generating code, [izer/izer.py calls the function load](https://github.com/MaximIntegratedAI/ai8x-synthesis/blob/bb712b97e8f27d24a51a787a3b78125460da52e1/izer/izer.py#L75) from izer/checkpoint.py, which [right shifts the bias values by 7](https://github.com/MaximIntegratedAI/ai8x-synthesis/blob/bb712b97e8f27d24a51a787a3b78125460da52e1/izer/checkpoint.py#L177) according to the BIAS_DIV constant [defined in izer/tornadocnn.py](https://github.com/MaximIntegratedAI/ai8x-synthesis/blob/bb712b97e8f27d24a51a787a3b78125460da52e1/izer/tornadocnn.py#L249). However, this procedure causes the biases to be effectively rounded to the weight bit precision, and these values (which are, e.g., {-2,-1,0,+1} for 2-bit) are transferred to the hardware as bias. Then, as far as I understand, these bias values are left shifted by 7 before being applied to (summed with) the convolution output, and then the sum is scaled down to get it back to the original order of magnitude (as per the leftmost part of the block diagram in the [User Guide](https://github.com/MaximIntegratedAI/MaximAI_Documentation/blob/aff49efec79b3c9b921f8cf0ec5c2babcb4de3fe/MAX78000/MAX78000%20User%20Guide.pdf), Figure 26-2. I think the right shift by 7 mentioned in the ["multiplication" section of the readme.md](https://github.com/MaximIntegratedAI/ai8x-synthesis/blob/bb712b97e8f27d24a51a787a3b78125460da52e1/README.md) also corresponds to this by the way. If not, I'm very confused, because then the bias values and the weight-activation multiplication result would be off by a factor of 128) 

**My question is:** I couldn't find anything on the User Guide or on any other piece of the documentation that would suggest that this is intended behavior, but is it intended behavior? By reading the User Guide, I can't see any reason for disallowing e.g., the use of 0x25, 0xd3 etc. (8-bit) bias values for a layer with 2-bit weights. Am I missing something? Because if this is intended, the following statement on the readme.md is not clear to me:

> On MAX78000/MAX78002, weights can be 1, 2, 4, or 8 bits wide... **Bias values are always 8 bits wide**. Data is 8 bits wide...

18.04.2021 edit: grammar

**Comment by rotx-maxim** [comment]
#126 (in develop) should fix this issue. Thanks for the great bug report!

**Comment by sonebu** [comment]
Yes, I just saw the merge. No problem, thanks for the fast response!
